core/src/plugins/filed/python/python-fd.cc

```diff
@@ -207,6 +221,104 @@ bRC unloadPlugin()
  * plugin instance must be thread safe and keep its own
  * local data.
  */
+
+// we use a vector instead of a set here since we expect that each thread
+// only accesses very few interpreters (<= 1) at the same time.
+thread_local std::vector<PyThreadState*> tl_threadstates{};
+
+/* Return this threads thread state for interp if it exists.  Returns
+ * nullptr otherwise */
+static PyThreadState* GetThreadStateForInterp(PyInterpreterState* interp)
+{
+  for (auto* thread : tl_threadstates) {
+    if (thread->interp == interp) { return thread; }
+  }
+  return nullptr;
+}
+
+static PyThreadState* PopThreadStateForInterp(PyInterpreterState* interp)
+{
+  auto iter = std::find_if(
+      tl_threadstates.begin(), tl_threadstates.end(),
+      [interp](const auto& thread) { return thread->interp == interp; });
+
+  if (iter != tl_threadstates.end()) {
+    auto* thread = *iter;
+
+    tl_threadstates.erase(iter);
+
+    return thread;
+  } else {
+    return nullptr;
+  }
+}
+
+class locked_threadstate {
+ private:
+  locked_threadstate(PyThreadState* ts, bool owns) : ts{ts}, owns{owns}
+  {
+    // lock the gil and make ts active
+    PyEval_RestoreThread(ts);
+  }
+
+ public:
+  explicit locked_threadstate(PyThreadState* ts) : locked_threadstate(ts, false)
+  {
+  }
+
+  explicit locked_threadstate(PyInterpreterState* interp)
+      : locked_threadstate(PyThreadState_New(interp), true)
+  {
+  }
+
+  locked_threadstate(const locked_threadstate&) = delete;
+  locked_threadstate& operator=(const locked_threadstate&) = delete;
+
+  locked_threadstate(locked_threadstate&& other) { *this = std::move(other); }
+
+  locked_threadstate& operator=(locked_threadstate&& other)
+  {
+    std::swap(ts, other.ts);
+    std::swap(owns, other.owns);
+
+    return *this;
+  }
+
+  PyThreadState* get() { return ts; }
+
+  ~locked_threadstate()
+  {
+    if (ts) {
+      if (owns) {
+        // destroy the thread state and release the gil
+        PyThreadState_Clear(ts);  // required before delete
+        PyThreadState_DeleteCurrent();
+      } else {
+        // just release the gil and make ts inactive
+        PyEval_ReleaseThread(ts);
+      }
+    }
+  }
+
+ private:
+  PyThreadState* ts{nullptr};
+  bool owns{false};
+};
+
+/* Acquire the gil for this thread.  If this thread does not have a thread
+ * state for interp, a new one is created.  This newly created thread state
+ * is destroyed by locked_threadstates destructor. */
+static locked_threadstate AcquireLock(PyInterpreterState* interp)
+{
+  auto* ts = GetThreadStateForInterp(interp);
+  if (!ts) {
+    // create a new thread state
+    return locked_threadstate{interp};
+  }
+
+  return locked_threadstate{ts};
+}
```

Comment:
I totally see how this works for threads created on the C++ side. However, does this also work for a thread created on the Python side?

Reply:
Have not tested it. So we are in the situation where a python-created thread tries to call some bareos api function ?

Reply:
Correct my if i am wrong, but i think these functions can only be called from the C++ side. python functions instead call to the bareos{fd,sd,dir}.cc functions